### ISSUE TITLE
How about making _|_ a record instead of a datatype?

### ISSUE THREAD

**Issue by gallais** [agda/agda-stdlib#192]
The notion of [divisibility](https://github.com/agda/agda-stdlib/blob/master/src/Data/Nat/Divisibility.agda#L33) is defined using a datatype which makes it
impossible to pattern match against a proof that `m | n` in a let-binding.

I suggest using the following record-based definition instead:

```agda
record _∣_ (m n : ℕ) : Set where
  constructor divides
  field quotient : ℕ
        equality : n ≡ quotient * m
```

However this will break some proofs: for some reasons `m` and `n` were not
declared as parameters of `_|_` but rather as indices. This means that the
`divides` constructor was carrying them as implicit arguments and this is used
in the standard library (see e.g. [∣-antisym](https://github.com/agda/agda-stdlib/blob/master/src/Data/Nat/Divisibility.agda#L64)).

Any opposition to the breaking change?

**Comment by asr** [comment]
I couldn't find any difference using both definitions (the data type and the record). Could you write-down a MWE?

**Comment by gallais** [comment]
The following definition only works if `m ∣ n` is defined as a record:

```agda
extract : {m n : ℕ} → m ∣ n → ℕ
extract pr = let (divides q _) = pr in q
```

**Comment by andreasabel** [comment]
The only reason I see for a data type here is printing.  But this should not stop us...

**Comment by MatthewDaggitt** [comment]
Closed with eeb0690
